### DISCUSSION TITLE
Accessing process.env in component

### DISCUSSION THREAD

**Discussion by ghost** [remix-run/remix#2928] [Q&A] [answered]
Actually is there a way to access all env variable inside the component, I can access to `process.env.NODE_ENV` like below:
```tsx
function TestComponent() {
  return (
    <div>{process.env.NODE_ENV}</div>
  )
}
```
but getting `process is not defined` error in browser console when i try to make component access to custom env variable like below:
```tsx
function TestComponent() {
  return (
    <div>{process.env.DOMAIN === "example.com" ? "Main Website Domain" : "Not Main Website Domain"}</div>
  )
}
```

**Comment by sergiodxa** [comment, marked as answer]
`process.env` is a Node specific thing, so that only works server side, you can read env variables inside a loader and return them as part of the data the route needs, then use `useLoaderData` to access the value.

**Comment by ghost** [comment]
but why it can access `process.env.NODE_ENV`?

**Reply by sergiodxa** [reply]
NODE_ENV is a special case, because many libraries, like React, uses that to do different things in dev or production (e.g. show more descriptive errors in dev and optimize the build for production) Remix replace `process.env.NODE_ENV` with the value of the NODE_ENV variable at build time, this means NODE_ENV is a build variable and not an env variables like the rest of the variables you could access using `process.env`.

**Comment by pmq24** [comment]
Can we have some ways to expose public env variables?

Example: `BASE_URL`. My scenario is the logo. The logo appears on every page and requires the env variable `BASE_URL` to redirect users to the homepage. This means I have to define the loader function for every page to expose the `BASE_URL` for the logo.


Maybe we can do the way of React (prefix `PUBLIC_` to the variable)?

**Reply by sergiodxa** [reply]
You can return the variable from your root loader and use an inline script to make it available on `window.process`, I wrote about this a few weeks ago here https://sergiodxa.com/articles/use-process-env-client-side-with-remix

This way you have full control of what variables you're making available to the client, and they're actual runtime env variables, meaning you can change them without doing a new build.

**Reply by pmq24** [reply]
That's a very good solution, thank you!

**Reply by dusty** [reply]
Here is what I do if you don't to use window for whatever reason.

root.tsx
```
export function loader() {
  const ENV = {
    THING: process.env.THING ?? ''
    STUFF: process.env.STUFF ?? ''
  }
  return json({otherThings, ENV})
}
export type LoaderData = SerializeFrom<typeof loader>
```

use-env.tsx
```
import { useMatches } from '@remix-run/react'
import type { LoaderData } from '~/root'

export const useEnv = () => {
  const matches = useMatches()
  const { ENV } = (matches.find((route) => {
    return route.id === 'root'
  })?.data || {}) as LoaderData
  return ENV || {}
}
```

Some component somewhere
```
export default function MyComponent() {
  const ENV = useEnv()
  
  return (
    <div>{ENV.THING === "example.com" ? "Main Website Domain" : "Not Main Website Domain"}</div>
  )
}
```

**Reply by pmq24** [reply]
Thanks a lot, you guys were very helpful!

**Reply by Ashutosh-Bhadauriya** [reply]
> You can return the variable from your root loader and use an inline script to make it available on window.process, I wrote about this a few weeks ago here https://sergiodxa.com/articles/use-process-env-client-side-with-remix

I tried this way but i keep getting reference error: window is not defined. Any help please?

**Reply by sergiodxa** [reply]
The reason to save it as `window.process.env` is that you can just do `process.env`, server-side it will access the actual process.env, client-side it will access `window.process.env`, but you shouldn't reference `window` directly because that doesn't exists server-side.

**Reply by cjoecker** [reply]
@sergiodxa do you know if this is still working on Remix after version `2.10.3` and vite with node `22.5.1`?

Before updating my dependencies, I could access the env variables in the server and the client with `process.env` if I had this as you describe it in one of your blog posts:
```html
			<script
				dangerouslySetInnerHTML={{
					__html: `window.process = ${JSON.stringify({
						env: data.env,
					})}`,
				}}
			/>
```

Now, I must access the variables as `window.env` in the client. 

I was sharing some configuration data between my server and client code as `{configVar: process.env.CONFIG_VAR}`. Now, I need to do it like this: `{configVar: process.env.CONFIG_VAR ?? window.env.CONFIG_VAR}` :(

**Comment by DeniCarvalho** [comment]
An alternative for me was to create an API Route to call an external api (BASE_URL) with axios and return the desired json to the component, using the fetch api [Remix doc](https://remix.run/docs/en/v1/guides/api-routes)

**Comment by aprilmintacpineda** [comment]
```
<script
  dangerouslySetInnerHTML={{
    __html: `window.process = ${JSON.stringify({
      env: data.ENV,
    })}`,
  }}
/>
```

This looks like a bad solution, it can potentially result to unwanted leaks of `process.env` variables. I think the best solution is for remix to inline all other process.env with `PUBLIC_` prefix to be accessible in the browser as well.

**Reply by kiliman** [reply]
`data.ENV` contains only ENV variables _you_ return from your root loader. So it's no less secure than adding a `PUBLIC_` prefix.

**Reply by sergiodxa** [reply]
Unless you return the whole `process.env`, but that's another thing, you're directly exposing your env variables in that case.

You could also not set them globally and instead always return them from the loader.

**Reply by aprilmintacpineda** [reply]
I think you missed the point. Remix has control over the generated javascript files, by introducing a `PUBLIC_` prefix to env variables, this kind of workaround can be prevented because there's always that lingering possibility of someone accidentally leaking a private environment variable. `data.ENV contains only ENV variables you return from your root loader. ` this does not prevent that possibility, a matter of fact, this is what makes it possible. But if remix will inject the environment variable with `PUBLIC` prefix to the generated javascript for the frontend, then I believe that possibility is completely eliminated, and remix should prevent env variables without `PUBLIC` prefix from reaching the frontend. Next and Vite are already doing this.

- https://vitejs.dev/guide/env-and-mode
- https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables#bundling-environment-variables-for-the-browser

## For a much safer way of implementing this workaround, here's what I did:

I created `serverDotenv.ts` and `browserDotenv.ts` files, this way I can separate env variable for the server and for the frontend despite them being defined in the same `.env` file

**serverDotenv.ts**
```ts
import * as yup from 'yup';

const serverEnvValidationSchema = yup.object({
  API_URL: yup.string().required(),
  API_KEY: yup.string().required()
});

const serverDotenv = serverEnvValidationSchema.validateSync(
  process.env,
  {
    stripUnknown: true
  }
);

export default serverDotenv;
```

**browserDotenv.ts**
```ts
import * as yup from 'yup';

const browserEnvValidationSchema = yup.object({
  PUBLIC_IMAGES_BASE_URL: yup.string().required()
});

const browserDotenv = browserEnvValidationSchema.validateSync(
  process.env,
  {
    stripUnknown: true
  }
);

export default browserDotenv;
```

Notice that I added `PUBLIC_` prefix to the environment variables that I intend to be injected to the browser, doing this allows me to add an additional check on the `root.tsx`. the `stripUnknown` option for yup is also important, this will remove other contents of `process.env` that was not defined in the schema validation. As an added bonus, this will also alert you if you forgot to add something to your `.env` file ahead of time without needing to call the endpoints that use them.

**root.tsx**
```ts
// ...other imports
import browserDotenv from './utils/browserDotenv';

export function loader() {
  return {
    ENV: browserDotenv
  };
}

export default function App () {
  // .. other codes

  useEffect(() => {
    Object.keys(process.env).forEach(envVariable => {
      // ensure that only env variables with PUBLIC_ makes it to the frontend
      if (!envVariable.startsWith('PUBLIC_')) {
        alert(
          `possible leakage of ${envVariable}. All env variables for the frontend should have PUBLIC_ prefix`
        );
      }
    });
  }, []);

  return (
    <>
      {/* other codes */}
      <script
        dangerouslySetInnerHTML={{
          __html: `window.process = ${JSON.stringify({
            env: data.ENV,
          })}`,
        }}
      />
    </>
  );
}
```
